New package: freerdp3-3.5.1

chrysos349 opened the pull request:
#### Testing the changes
- I tested the changes in this PR: **YES**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

#### Local build testing
- I built this PR locally for my native architecture, (x86_64)
- I built this PR locally for these architectures (if supported. mark crossbuilds):
  - aarch64-musl x
  - armv7l x
  - armv6l-musl x

ahesford commented on srcpkgs/cJSON/template:
```diff
@@ -1,14 +1,20 @@
-checksum=5308fd4bd90cef7aa060558514de6a1a4a0819974a26e6ed13973c5f624c24b2
+checksum=c91d1eeb7175c50d49f6ba2a25e69b46bd05cffb798382c19bfb202e467ec51c
+
+# help cmake find headers when cross-compiling
+post_patch() {
+	vsed -i CMakeLists.txt library_config/cJSONConfig.cmake.in \
+		-e "/CMAKE_INSTALL/s/_FULL//g"
+}
```
Don't use `vsed` for this; add a proper patch.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+checksum=03323b383980ee91decbed88270bac061ffb17fd04e52576c70da7885601ecbe
+CFLAGS="-Wno-dev"
+
+# help cmake generate manuals when cross-compiling
+post_patch() {
+	vsed -i cmake/InstallFreeRDPMan.cmake \
+		-e '/generate_argument_docbook/s/COMMAND/& \${CMAKE_CROSSCOMPILING_EMULATOR}/'
+}
```
Don't use `vsed` here; add a proper patch.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+		-e '/generate_argument_docbook/s/COMMAND/& \${CMAKE_CROSSCOMPILING_EMULATOR}/'
+}
+
+pre_configure() {
+	if [ "$CROSS_BUILD" ]; then
+		export CXX="$XBPS_WRAPPERDIR/${XBPS_CROSS_TRIPLET}-c++"
+		configure_args+="
+		 -DCMAKE_CXX_COMPILER_AR=/usr/bin/${XBPS_CROSS_TRIPLET}-ar
+		 -DCMAKE_CXX_COMPILER_RANLIB=/usr/bin/${XBPS_CROSS_TRIPLET}-ranlib"
+	fi
+}
```
`$CXX` is usually defined to `${XBPS_CROSS_TRIPLET}-c++` automatically, and the wrapper directory should take precedence in the path. Is the first line really required? Also, the `configure_args` appends should be outside of a function.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+	fi
+}
+
+case "$XBPS_TARGET_MACHINE" in
+	i686*|x86_64*) configure_args+=" -DWITH_SSE2=ON";;
+	armv5tel*) configure_args+=" -DWITH_NEON=OFF";;
+esac
+
+# first we want to separate -server, everything else then goes to client
+subpackages="libfreerdp-server3 freerdp-server3 freerdp-server3-devel
+ libfreerdp3 freerdp3-devel"
```
These top-level blocks belong above all function definitions.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+libfreerdp-server3_package() {
```
Package should be called `libfreerdp3-server`.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+freerdp-server3-devel_package() {
+	depends="libfreerdp-server3>=${version}_${revision}
+	 freerdp3-devel>=${version}_${revision}"
+	short_desc="Free RDP (Remote Desktop Protocol) server - development files"
+	pkg_install() {
+		vmove "usr/lib/libfreerdp-server*.so"
+		vmove "usr/lib/libfreerdp-shadow*.so"
+		vmove usr/lib/cmake/FreeRDP-Proxy3
+		vmove usr/lib/cmake/FreeRDP-Server3
+		vmove usr/lib/cmake/FreeRDP-Shadow3
+		vmove usr/lib/pkgconfig/freerdp-server-proxy3.pc
+		vmove usr/lib/pkgconfig/freerdp-server3.pc
+		vmove usr/lib/pkgconfig/freerdp-shadow3.pc
+	}
+}
```
Just lump this in with `freerdp3-devel`; there isn't much value in separating server development files.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+freerdp-server3_package() {
```
Package should be called `freerdp3-server`.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+}
+
+libfreerdp-server3_package() {
+	short_desc="Free RDP (Remote Desktop Protocol) server - runtime files"
```
```suggestion
	short_desc="Free RDP (Remote Desktop Protocol) server - runtime libraries"
```

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+
+freerdp-server3_package() {
+	short_desc="Free RDP (Remote Desktop Protocol) server"
+	depends="libfreerdp-server3>=${version}_${revision}"
```
This ought to be unnecessary, because shlib dependencies should pull it in.

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,143 @@
+
+libfreerdp-server3_package() {
+	short_desc="Free RDP (Remote Desktop Protocol) server - runtime files"
+	depends="libfreerdp3>=${version}_${revision}"
```
I assume that the libraries in this package will pull in `libfreerdp3` automatically with shlib dependencies. If so, you can drop this line.

ahesford commented:
In the future, please refrain from opening competing PRs. Offer guidance on existing PRs and give the original author a chance to address outstanding issues. Multiple competing requests just adds noise and can be frustrating for contributors.

fanyx commented:
as default dependency [client/SDL/aad/CMakeLists.txt](https://github.com/FreeRDP/FreeRDP/blob/1687a6424a4b4f43a1698c0c2d850917bd37c92b/client/SDL/aad/CMakeLists.txt#L47) asks for `webkit2gtk-4.0`

looking at their own pipelines and arch's repo we can probably import [0001-Use-webkit2gtk-4.1.patch](https://gitlab.archlinux.org/archlinux/packaging/packages/freerdp/-/blob/main/0001-Use-webkit2gtk-4.1.patch) to patch to webkit2gtk-4.1 and use `libwebkit2gtk41`

chrysos349 commented:
@ahesford 
thanks for reviewing the template. fixed everything.
regarding "needlessly" exporting `CXX`, it is actually needed, or else the crossbuild will fail.

> In the future, please refrain from opening competing PRs

i'll keep it in mind.

@fanyx
being an optional dep, sdl2 pulls webkit which in turn pull a ton of packages. i decided to skip it (ubuntu does, for example). 
otherwise, `freerdp3` will pull lots of deps.
p.s. another option is to split sdl version into a separate subpackage.
p.p.s. if it were up to me, i would also split x11 and wayland versions into separate packages also (debian does so).

ahesford commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,119 @@
+ -DWITH_CAIRO=ON
+ -DWITH_CHANNELS=ON
+ -DWITH_CLIENT_CHANNELS=ON
+ -DWITH_CLIENT_SDL=ON
```
This either has to be dropped, or `makedepends` needs `SDL2-devel`, `SDL2_ttf-devel` and `webkit2gtk-devel`, although I can't get `sdl-freerdp3` to actually show me a remote window.

chrysos349 replied:
enabled sdl client without webview.

ahesford commented:
`wlfreerdp3` says it's deprecated, and it's also segfaulting whenever I run it:

```
Thread 1 "wlfreerdp3" received signal SIGSEGV, Segmentation fault.
__memcpy_avx_unaligned_erms ()
    at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:488
warning: 488	../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory
(gdb) bt
#0  __memcpy_avx_unaligned_erms ()
    at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:488
#1  0x000055555555cb72 in memcpy (__len=<optimized out>, __src=0x7fffdfbb8030, 
    __dest=<optimized out>) at /usr/include/bits/string_fortified.h:29
#2  wlf_copy_image (area=<optimized out>, scale=0, dstHeight=900, dstWidth=1440, 
    dstStride=5760, dst=0x1, srcHeight=<optimized out>, srcWidth=<optimized out>, 
    srcStride=5760, src=<optimized out>) at ../client/Wayland/wlfreerdp.c:765
#3  wlf_copy_image (src=<optimized out>, srcStride=5760, srcWidth=<optimized out>, 
    srcHeight=<optimized out>, dst=dst@entry=0x7fffdece2000, 
    dstStride=dstStride@entry=5760, dstWidth=1440, dstHeight=900, area=0x7fffffffdc98, 
    scale=0) at ../client/Wayland/wlfreerdp.c:738
#4  0x0000555555563925 in wl_update_buffer (context_w=0x55555559cdc0, ix=0, 
    iy=<optimized out>, iw=1440, ih=900) at ../client/Wayland/wlfreerdp.c:98
#5  0x00007ffff7d25fec in freerdp_connect (instance=instance@entry=0x55555559cae0)
    at ../libfreerdp/core/freerdp.c:208
#6  0x000055555555acab in wlfreerdp_run (instance=0x55555559cae0)
    at ../client/Wayland/wlfreerdp.c:491
#7  main (argc=9, argv=0x7fffffffe238) at ../client/Wayland/wlfreerdp.c:728
```

Not sure if we should just drop Wayland, or try to fix this issue.

chrysos349 commented:
disabled wlfreerdp3 client (deprecated), enabled sdl-freerdp3 client (experimental), disabled webview.

sdl-freerdp3 works for me: show the connection dialog, and connect to the remote box.

oreo639 commented on srcpkgs/freerdp3/template:
```diff
@@ -0,0 +1,137 @@
+revision=1
+build_style=cmake
+build_helper=qemu
+configure_args="
```
Can you groups these so that multiple options share a line when possible.
